Coverage/FileInfo.h

```diff
@@ -3,48 +3,112 @@
 #include "FileLineInfo.h"
 #include "RuntimeOptions.h"
 
+#include <algorithm>
 #include <vector>
 #include <string>
 #include <iostream>
 #include <fstream>
 
 struct FileInfo
 {
-	FileInfo(const std::string& filename)
+private:
+	enum class LineType
 	{
-		std::ifstream ifs(filename);
+		CODE,
+		ENABLE_COVERAGE,
+		DISABLE_COVERAGE
+	};
 
-		bool current = true;
+	static constexpr std::string_view DISABLE_COVERAGE = "DisableCodeCoverage";
+	static constexpr std::string_view ENABLE_COVERAGE = "EnableCodeCoverage";
+	static constexpr std::string_view PRAGMA_LINE = "#pragma";
+	static constexpr std::string_view DOUBLE_FORWARD_SLASH_LINE = "//";
+	static constexpr std::string_view FORWARD_SLASH_ASTERISK_LINE = "/*";
+	static constexpr std::string_view FORWARD_SLASH_ASTERISK_LINE_END = "*/";
 
-		std::string line;
-		while (std::getline(ifs, line))
+	bool IsCoverageFlag(const std::string::const_iterator& iter, const ptrdiff_t iterSize,
+											const std::string_view& coverageFlag, const bool multilineComment)
+	{
+		if (iterSize < coverageFlag.size())
 		{
-			// Process str
-			size_t idx = line.find("#pragma");
+			return false;
+		}
+		if (iterSize == coverageFlag.size())
+		{
+			return std::equal(coverageFlag.begin(), coverageFlag.end(), iter);
+		}
+
+		if (!multilineComment)
+		{
+			return false;
+		}
+
+		if (!std::equal(coverageFlag.begin(), coverageFlag.end(), iter))
+		{
+			return false;
+		}
+
+		const ptrdiff_t restSize = iterSize - coverageFlag.size();
+		if (restSize < FORWARD_SLASH_ASTERISK_LINE_END.size()) {
+			return false;
+		}
+
+		return std::equal(FORWARD_SLASH_ASTERISK_LINE_END.begin(),
+											FORWARD_SLASH_ASTERISK_LINE_END.end(),
+											iter + coverageFlag.size());
+	}
+
+	LineType GetLineType(const std::string& line)
+	{
+		static const std::vector<std::string_view> prefixCoverage{ PRAGMA_LINE, DOUBLE_FORWARD_SLASH_LINE, FORWARD_SLASH_ASTERISK_LINE };
```

Comment:
Not all prefixes are the same. Comments have a begin (and optional end), while pragma's need to be at the start of a line. 

I'm also wondering what would happen when we get to a mixture of `#pragma` and then end that section with a comment. Especially with nesting? I guess that's allowed and it'll pick the latest (no nesting) which I suppose is fine.

Reply:
Especially multi-line comments are nasty to handle correctly I think. I can already see people using it like:

```c++
    int foo = a + 2 * /* DisableCodeCoverage */ myMethod(); /* EnableCodeCoverage */
```

... that is: I could see many reasons to do this. There is a case for using code coverage like this, even though I don't think it should be supported.